spec/mysql2/client_spec.rb

```diff
@@ -461,79 +452,56 @@ def connect *args
         }.should raise_error(Mysql2::Error)
       end
 
-      it "should close the connection when an exception is raised" do
-        begin
-          Timeout.timeout(1, Timeout::Error) do
-            @client.query("SELECT sleep(2)")
-          end
-        rescue Timeout::Error
-        end
+      if Thread.respond_to?(:handle_interrupt)
+        it 'should be impervious to connection-corrupting timeouts ' do
+          # attempt to break the connection
+          expect { Timeout.timeout(0.1) { @client.query('SELECT SLEEP(1)') } }.to raise_error(Timeout::Error)
 
-        lambda {
-          @client.query("SELECT 1")
-        }.should raise_error(Mysql2::Error, 'closed MySQL connection')
+          # expect the connection to not be broken
+          expect { @client.query('SELECT 1') }.to_not raise_error
+        end
       end
 
-      it "should handle Timeouts without leaving the connection hanging if reconnect is true" do
-        client = Mysql2::Client.new(DatabaseCredentials['root'].merge(:reconnect => true))
-        begin
-          Timeout.timeout(1, Timeout::Error) do
-            client.query("SELECT sleep(2)")
-          end
-        rescue Timeout::Error
+      context 'when a non-standard exception class is raised' do
```

Comment:
@sodabrew i think this is the real behaviour we want to demonstrate and test for - we will protect the connection from regular `Timeout.timeout` but not weird arbitrary exceptions